cloud/ec2.go

```diff
@@ -167,6 +172,130 @@ func (m *ec2Manager) Configure(ctx context.Context, settings *evergreen.Settings
+	header.Add("Content-Type", contentType)
+	header.Add("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", fileName))
+
+	part, err := writer.CreatePart(header)
+	if err != nil {
+		return errors.Wrap(err, "error making custom user data part")
+	}
+
+	if _, err := part.Write([]byte(userDataPart)); err != nil {
+		return errors.Wrap(err, "error writing custom user data")
+	}
+
+	return nil
+}
+
+var userDataPrefixToContentType = map[string]string{
+	"#!":              "text/x-shellscript",
+	"#include":        "text/x-include-url",
+	"#cloud-config":   "text/cloud-config",
+	"#upstart-job":    "text/upstart-job",
+	"#cloud-boothook": "text/cloud-boothook",
+	"#part-handler":   "text/part-handler",
+	"<powershell>":    "text/x-shellscript",
+	"<script>":        "text/x-shellscript",
+}
+
+// userDataContentType detects the content type based on the directive on the
+// first line of the user data.
+func userDataContentType(userData string) (string, error) {
+	var firstLine string
+	index := strings.IndexByte(userData, '\n')
+	if index == -1 {
+		firstLine = userData
+	} else {
+		firstLine = userData[:index]
+	}
+	firstLine = strings.TrimSpace(firstLine)
+
+	for key, val := range userDataPrefixToContentType {
+		if strings.HasPrefix(firstLine, key) {
+			return val, nil
+		}
+	}
+	return "", errors.Errorf("user data format is not recognized from first line: '%s'", firstLine)
+}
+
+// makeMultipartUserData returns user data in a multipart MIME format with the
+// given files and their content.
+func makeMultipartUserData(files map[string]string) (string, error) {
+	buf := &bytes.Buffer{}
+	parts := multipart.NewWriter(buf)
+
+	if err := writeUserDataHeaders(buf, parts.Boundary()); err != nil {
+		return "", errors.Wrap(err, "error writing MIME headers")
+	}
+
+	for fileName, content := range files {
+		if err := writeUserDataPart(parts, content, fileName); err != nil {
+			return "", errors.Wrapf(err, "error writing user data '%s'", fileName)
+		}
+	}
+
+	if err := parts.Close(); err != nil {
+		return "", errors.Wrap(err, "error closing MIME writer")
+	}
+
+	return buf.String(), nil
+}
+
+// bootstrapScript creates the user data script that bootstraps the host.
+func bootstrapScript(h *host.Host, config evergreen.JasperConfig, dir string) string {
+	if h.Distro.IsWindows() {
+		cmds := h.FetchJasperCommandWithPath(config, dir, "/bin")
+		// PowerShell nested quotation marks are handled by using two quotation
+		// marks.
+		quotedCmds := strings.Replace(cmds, "'", "''", -1)
+		commands := []string{
+			"<powershell>",
+			fmt.Sprintf("C:\\cygwin\\bin\\bash.exe -c '%s'", quotedCmds),
+			"</powershell>",
+		}
+		return strings.Join(commands, "\r\n")
+	}
+	return strings.Join([]string{"#!/bin/bash", h.FetchJasperCommand(config, dir)}, "\n")
+}
+
 func (m *ec2Manager) spawnOnDemandHost(ctx context.Context, h *host.Host, ec2Settings *EC2ProviderSettings, blockDevices []*ec2.BlockDeviceMapping) ([]*string, error) {
```

Comment:
userdata is a think that's available in some way, for every cloud provider... while there might be some means to control how bootstrapping works for a cloud provider, in general, most of this code should happen above the manager implementation.